### DISCUSSION TITLE
StreamLostError when idle for a few hours

### DISCUSSION THREAD

**Discussion by zhiyan114** [pika/pika#1428] [Q&A] [answered]
I was able to solve the `Stream connection lost: ConnectionResetError(104, 'Connection reset by peer')` by using Threading for tasks that takes well over a minute to process, but somehow this same errors shows up after a few hours of idling.

I'm very new to Threading so I'm wondering if I had somehow misused it that caused this or it could just be a general network issue?

Main.py
```py
def cbAck(ch, delivery_tag):
    if ch.is_open:
        ch.basic_ack(delivery_tag)
def cbSend(jsonData):
    queueHandler.sendToQueue(jsonData)

def callback(ch, method, properties, body, conn):
    data = json.loads(body.decode("utf-8"))
    print(data["interactID"]+": Processing for user "+data["userID"]+"...", flush=True)
    # Download the audio file and start processing it
    fileName = SaveFileToDisk(data["mediaLink"])
    if fileName is None:
        print(data["interactID"]+": Failed to download file", flush=True)
        return conn.add_callback_threadsafe(functools.partial(cbSend, json.dumps({
            "success": False,
            "userID": data["userID"],
            "interactID": data["interactID"],
            "cost": data["cost"],
            "reason": "Failed to download file"
        })))
    print(data["interactID"]+": File downloaded", flush=True)
    # Process the file
    start = time.time()
    result = inference.convert(fileName, data["language"])
    end = time.time()
    os.remove(fileName)
    print(data["interactID"]+": Processed in "+str(end-start)+" seconds", flush=True)
    # Send the result back
    conn.add_callback_threadsafe(functools.partial(cbSend, json.dumps({
        "success": True,
        "userID": data["userID"],
        "interactID": data["interactID"],
        "cost": data["cost"],
        "result": result['text'],
        "processTime": end-start
    })))
    # Acknowledge the message
    conn.add_callback_threadsafe(functools.partial(cbAck, ch, method.delivery_tag))
    #ch.basic_ack(delivery_tag = method.delivery_tag)




if __name__ == '__main__':
    try:
        print("Starting OpenAI Whisper ML Server...")
        queueHandler.receiveFromQueue(callback)
    except KeyboardInterrupt:
        print("Shutting down OpenAI Whisper ML Server...")
        try:
            sys.exit(0)
        except SystemExit:
            os._exit(0)
```

queueHandler.py
```py
import pika
import os
import ssl
import functools
import threading


sendQName = "WhisperRes"
receiveQName = "WhisperReq"

def strtobool (val):
    val = val.lower()
    if val in ('y', 'yes', 't', 'true', 'on', '1'):
        return True
    elif val in ('n', 'no', 'f', 'false', 'off', '0'):
        return False
    else:
        raise ValueError("invalid truth value %r" % (val,))


pikaCred = pika.PlainCredentials(os.environ.get('AMQP_USER', 'guest'), os.environ.get('AMQP_PASS', 'guest'))
sslContext = ssl.create_default_context()
pikaParams = pika.ConnectionParameters(
    host=os.environ.get('AMQP_HOST', 'localhost'),
    port=int(os.environ.get('AMQP_PORT', "5672")),
    virtual_host=os.environ.get("AMQP_VHOST", "/"),
    credentials=pikaCred,
    ssl_options=pika.SSLOptions(sslContext) if strtobool(os.environ.get("AMQP_TLS", "false")) else None,
    heartbeat=60
)

connection = pika.BlockingConnection(pikaParams)
sendChannel = connection.channel()
receiveChannel = connection.channel()
receiveChannel.basic_qos(prefetch_count=1) # Only receive one message at a time

def sendToQueue(message):
    sendChannel.queue_declare(queue=sendQName, durable=True)
    sendChannel.basic_publish(exchange='', routing_key=sendQName, body=message)
def receiveFromQueue(callback):
    def internal_callback(ch, method, properties, body):
        threading.Thread(target=callback, args=(ch,method,properties,body,connection)).start()
    receiveChannel.queue_declare(queue=receiveQName, durable=True)
    receiveChannel.basic_consume(queue=receiveQName, on_message_callback=internal_callback, auto_ack=False) # Acknowledge the message after processing
    receiveChannel.start_consuming()
```

**Comment by shinji-s** [comment]
Could the cause be unstable execution order of cbAck and cbSend? I'm not sure if pika makes any guarantee on the execution order of registered callbacks.

**Reply by shinji-s** [reply]
Please disregard above reply. I mistook the second callback be a channel closure.

**Reply by zhiyan114** [reply]
Thanks for the reply, I just tested it with complete idle last night (i.e. no message from the start) and was able to reproduce the same error after a few hours (this time, it lasted a little longer).

I'm starting to suspect that it could be some other issues.

**Comment by lukebakken** [comment]
I need to know what RabbitMQ logs at the same time as the `ConnectionResetError`.

**Reply by zhiyan114** [reply]
I'm currently using cloudamqp free plan, which doesn't have the log option. Let me spin-up a rabbitMQ instance and try to replicate the error on that server.

**Reply by zhiyan114** [reply]
The crash occurred little earlier than usual, here's what I gather from the log:

```
2023-04-10 12:58:54.681025+00:00 [info] <0.16494.0> accepting AMQP connection <0.16494.0> ([CLI_IP]:52702 -> [SRV_IP]:5671)
2023-04-10 12:58:54.737729+00:00 [info] <0.16494.0> connection <0.16494.0> ([CLI_IP]:52702 -> [SRV_IP]:5671): user 'USERNAME' authenticated and granted access to vhost 'VHOSTNAME'
[...]
2023-04-10 16:30:55.187593+00:00 [warning] <0.16494.0> closing AMQP connection <0.16494.0> ([CLI_IP]:52702 -> [SRV_IP]:5671, vhost: 'VHOSTNAME', user: 'USERNAME'):
2023-04-10 16:30:55.187593+00:00 [warning] <0.16494.0> client unexpectedly closed TCP connection
```

The log doesn't seem to be useful as it shows the disconnection was the client's fault. I do have the crash trace logged on sentry.io if that'll help.

**Reply by lukebakken** [reply]
Sure, the client logs would be good. If you can run the client with DEBUG logging that could be interesting - https://github.com/pika/pika/blob/main/examples/basic_consumer_threaded.py#L11-L15

Do you know what sort of network devices are between your client and RabbitMQ? Load balancers, firewalls, etc? What cloud provider are you using?

The code as-is looks fine.

**Reply by zhiyan114** [reply]
Great, I'll just try to just log around and see what I can find. Let me know if you want me to log anything specific. I'm also using cloudamqp for the main server and Oracle for the test server.

**Reply by zhiyan114** [reply]
Just figured it out. Apparently my router (or ISP) randomly started to randomly drop packets (which I notice when my game session got disconnected at the exact same time the crash happens).

Would this be a good way to handle reconnection?
```py
pikaConnDat = {
    "connection": None,
    "sendChannel": None,
    "receiveChannel": None
}
def init():
    pauseTime = 5
    while True:
        try:
            pikaConnDat['connection'] = pika.BlockingConnection(pikaParams)
            pikaConnDat['sendChannel'] = pikaConnDat['connection'].channel()
            pikaConnDat['sendChannel'].queue_declare(queue=sendQName, durable=True)
            pikaConnDat['receiveChannel'] = pikaConnDat['connection'].channel()
            pikaConnDat['receiveChannel'].queue_declare(queue=receiveQName, durable=True)
            pikaConnDat['receiveChannel'].basic_qos(prefetch_count=1) # Only receive one message at a time
            break
        except Exception:
            print(f"Failed to connect to RabbitMQ, retrying in {pauseTime} seconds...")
            time.sleep(pauseTime)
            if(pauseTime < 30):
                pauseTime += 5


def sendToQueue(message):
    pikaConnDat['sendChannel'].basic_publish(exchange='', routing_key=sendQName, body=message)
def receiveFromQueue(callback):
    def internal_callback(ch, method, properties, body):
        threading.Thread(target=callback, args=(ch, method, properties, body, pikaConnDat['connection'])).start()
    while True:
        pikaConnDat['receiveChannel'].basic_consume(queue=receiveQName, on_message_callback=internal_callback, auto_ack=False) # Acknowledge the message after processing
        try:
            pikaConnDat['receiveChannel'].start_consuming()
        except (pika.exceptions.StreamLostError, pika.exceptions.AMQPHeartbeatTimeout):
            print("Network dropped, reconnecting...")
            init()
            print("Reconnected")
```

**Reply by lukebakken** [reply, marked as answer]
Use this as a starting point - https://github.com/pika/pika/blob/main/examples/blocking_consume_recover_multiple_hosts.py

**Reply by zhiyan114** [reply]
Alright, thank you!

**Reply by rithvikshetty** [reply]
Did it help? I tried the same approach but my connection is still going down after 16-17 hours.

**Reply by lukebakken** [reply]
It would be **VERY HELPFUL** to know what RabbitMQ logs at the same time as these connection issues!

**Reply by zhiyan114** [reply]
> Did it help? I tried the same approach but my connection is still going down after 16-17 hours.

I just setup an algorithm to reconnect since I found out that it was my router that's randomly disconnecting for a few seconds.

**Reply by rithvikshetty** [reply]
@lukebakken @zhiyan114 
These are the logs:



ERROR:pika.adapters.utils.io_services_utils:_AsyncBaseTransport._consume() failed, aborting connection: error=ConnectionResetError(104, 'Connection reset by peer')

Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/pika/adapters/utils/io_services_utils.py", line 1047, in _on_socket_readable
    self._consume()
  File "/usr/local/lib/python3.11/site-packages/pika/adapters/utils/io_services_utils.py", line 791, in _consume
    data = self._sigint_safe_recv(self._sock, self._MAX_RECV_BYTES)

**Reply by lukebakken** [reply]
I specifically asked for what **RabbitMQ** logs when Pika is disconnected.

`Connection reset by peer` is probably caused by a network device between your application and RabbitMQ.

**Reply by alessandro-gandolfi-ors** [reply]
Hello, I'm having a similar problem with a component using Pika. It works for some time, maybe some hours, then it stops completely but I can't understand why.

Here is the stacktrace of the component using Pika:
```
Traceback (most recent call last):
  File "/app/main.py", line 406, in <module>
    main()
  File "/app/main.py", line 386, in main
    channel.basic_ack(
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 2113, in basic_ack
    self._flush_output()
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 1336, in _flush_output
    self._connection._flush_output(lambda: self.is_closed, *waiters)
  File "/usr/local/lib/python3.9/site-packages/pika/adapters/blocking_connection.py", line 522, in _flush_output
    raise self._closed_result.value.error
pika.exceptions.StreamLostError: Stream connection lost: ConnectionResetError(104, 'Connection reset by peer')
```

Here is the error logged by Rabbit:
```
[error] <0.4070.2> closing AMQP connection <0.4070.2> (10.1.41.121:57236 -> 10.1.41.113:5672):
[error] <0.4070.2> {writer,send_failed,{error,timeout}}
```

This is part of component that was written by another team, that decided to dequeue, process and ack the messages in bulk.
The processing could take even less than a minute, the ack should happen after the processing.
The initial bulks are processed without any problems for some hours, after a while the error starts to be thrown.
The error is thrown when trying to ack the bulk of messages processed.
I don't think the processing is done in another thread but I tried to make the period between heartbits higher, it doesn't change anything, the error is thrown anyway after less then a minute after receiving the bulk and processing it.

Do you have any idea on what could be the problem? Thanks in advance.

**Reply by lukebakken** [reply]
@alessandro-gandolfi-ors please start a new discussion with that information, thank you: https://github.com/pika/pika/discussions
